index.js

```diff
@@ -241,6 +239,51 @@ class Auth {
       callback(null, env.IS_CONTAINER_ENGINE);
     });
   }
+
+  sign (data, callback) {
```

Comment:
If I'm reading this right, it looks like the call to the `signBlob` API will happen in all cases, even those where the client was initialized with a keyfile.  I _think_ the desired behavior is that the `sign()` function would sign the blob locally if the keyfile was present, and fall-back to the `signBlob` API if it wasn't.  Currently in GCP the keyfile is not available for the default application credential, but is _required_ if you want to auth the client with an alternate identity.  In those cases I imagine the expectation is that it's that identity which is signing the blob

Reply:
Thanks for reviewing the code. If I'm on GCP, didn't provide a key file to `require('@google-cloud/storage')`, but want to create a signed URL from another service account, that would look like:

```js
var gcs = require('@google-cloud/storage')() // ADC
gcs.getBuckets() // ... use `gcs` as necessary

// When I need to create a signed URL from an alternate identity:
var gcsOtherServiceAccount = require('@google-cloud/storage')({
  keyFilename: 'other-service-account-key.json'
})
var file = gcsOtherServiceAccount.bucket('bucket').file('file')
file.createSignedUrl({...})
```

Let me know if I misunderstood the scenario you were describing.

> I think the desired behavior is that the sign() function would sign the blob locally if the keyfile was present, and fall-back to the signBlob API if it wasn't.

Can we verify that? Is the concern the perf expense of the API call? Rather than signing locally, letting the API take on the responsibility for correct signing seems like the best way to go.

Reply:
I'm concerned that APIs have an annoying habit of having things like quotas and rate limits.  It's partly about the latency increase, partly about adding a point of failure, and partly about a change in behavior that may not be expected.

I'm also secretly worried about the cross-function use case (where the bucket to which the signed URL will point is in a different project from the client creating the URL), but haven't thought through whether that's more of an issue here or not.

There is also the fact that the IAM API needs to be enabled for the project, which it may not be and this would "suddenly" cause existing apps to start breaking unexpectedly

Reply:
OK cool.  Thanks

Reply:
@stephenplusplus After you have this completed, can you orchestrate a release as well?

Reply:
Updated with local signing, PTAL.

@lukesneeringer we need a merge and release of https://github.com/google/google-auth-library-nodejs/pull/180 before I can do this one.